Support having no metadata file for HuggingFaceStorageReader

ankitageorge opened the pull request:
Summary: If there is only one safetensors file, we don't need users to have a metadata file and we can just construct it from the keys of that file. This is a use-case for some HuggingFace models, so adding support for it

Test Plan:
ensure existing tests pass
tested e2e in a notebook

Differential Revision: D72472490




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

joecummings reviewed (approved):
Cool! Should there be a test attached?

ankitageorge committed:
Support having no metadata file for HuggingFaceStorageReader (#150701)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150701

If there is only one safetensors file, we don't need users to have a metadata file and we can just construct it from the keys of that file. This is a use-case for some HuggingFace models, so adding support for it

Test Plan:
ensure existing tests pass
tested e2e in a notebook

Differential Revision: D72472490

lw commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -479,6 +479,9 @@ def exists(self, path: Union[str, os.PathLike]) -> bool: ...
     @abstractmethod
     def rm_file(self, path: Union[str, os.PathLike]) -> None: ...
 
+    @abstractmethod
+    def ls(self, path: Union[str, os.PathLike]) -> list[str]: ...
```
By adding an additional abstract method to this abstract base class, this PR broke all users who subclassed `FileSystemBase` since such subclasses need to implement all abstract methods. Could you find a fix or revert this PR? Thanks
